### PR TITLE
first contribute to compas geometry

### PR DESCRIPTION
**Pull request by nikeftekhar**
contributes regarding next topics:
**polyline**: tween, divide by count, rebuild, shorten
**line**: divide by count
**angles**: between planes
**extend**: lines, polylines
**intersections**: mesh-line, mesh-plane
**splits**: polyline-plane, mesh-plane

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/geometry/_core/angles.py [hunk 1]
```diff
@@ -352,6 +352,30 @@ def angles_points_xy(a, b, c, deg=False):
     v = subtract_vectors_xy(c, a)
     return angles_vectors_xy(u, v, deg)
 
+def angle_planes(a, b, deg=False):
+    r""" Compute the smallest angle between the two normal vectors of two planes.
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: the raw string prefix is only necessary if you use latex syntax in the docstring...

File: src/compas/geometry/_core/angles.py [hunk 2]
```diff
@@ -352,6 +352,30 @@ def angles_points_xy(a, b, c, deg=False):
...
+    float
+        The smallest angle in radians (in degrees if deg == True).
+        The angle is always positive.
+
+    Examples
+    --------
+    >>> angle_planes(plane_a, plane_b, True)
+
+    """
+
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: please remove unnecessary whitespace

File: src/compas/geometry/_core/angles.py [hunk 3]
```diff
@@ -352,6 +352,30 @@ def angles_points_xy(a, b, c, deg=False):
...
+    Returns
+    -------
+    float
+        The smallest angle in radians (in degrees if deg == True).
+        The angle is always positive.
+
+    Examples
+    --------
+    >>> angle_planes(plane_a, plane_b, True)
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py:378]: the example would be more meaningful (and doctest-able) if you actually compute the angle between two planes, such as XY and YZ and output the result...

File: src/compas/geometry/_core/angles.py [hunk 4]
```diff
@@ -352,6 +352,30 @@ def angles_points_xy(a, b, c, deg=False):
...
+
+    """
+
+    return angle_vectors(a.normal, b.normal, deg)
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: this will only work if the input parameters are `Plane` objects. however, the base functions should be compatible with COMPAS objects and Python objects. since `Plane` objects implement the iteration protocol, the following would also allow the plane to be provided as a tuple of point and normal.
```python
return angle_vectors(a[1], b[1], deg)
```

File: src/compas/geometry/_core/intersections.py [hunk 5]
```diff
@@ -382,6 +382,37 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
 
     return None
 
+def intersection_polyline_plane(polyline, plane, expected_number_of_intersections=2, tol=1e-6):
+    """ Calculates the intersection point of a plane with a polyline. 
+        returns a list of intersections 
+        by default it will allow two intersections. reduce expected_number_of_intersections to speed up.
```
Comment (tomvanmele) [inline src/compas/geometry/_core/intersections.py]: Please remove whitespace between the triple quotes and the summary. Please also remove the indents of the second and third line.

File: src/compas/geometry/_core/intersections.py [hunk 6]
```diff
@@ -382,6 +382,37 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        Default is ``1e-6``.
+
+    Returns
+    -------
+    intersection_pts : list of compas.geometry.Point
+        if there are intersection points, return point(s) in a list
+    """
+    intersection_pts = []
+    max_iter =0
+    for segment in polyline.lines:
```
Comment (tomvanmele) [inline src/compas/geometry/_core/intersections.py]: core functions should be compatible with COMPAS geometry objects as well as their pure Python representations. a polyline is basically a sequence of points. lines are given by consecutive pairs of points. the `Polyline` supports the iteration protocol such that you can write the following.
```python
for segment in pairwise(polyline):
```

**Review by tomvanmele — COMMENTED** [review]
for a first contribution this is quite elaborate. although the effort is much appreciated, perhaps it would be better to split this up into two PRs with a narrower scope. this way we can "practice" with the first PR and make the review process of the second one a bit easier. i would suggest limiting the first PR to the angles between planes and the new intersection functions in `_core/intersections.py`...

File: src/compas/geometry/_core/angles.py (hunk 1 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/angles.py]: oh! i didn't know that. good to know, copied. thank you very much for the comment

File: src/compas/geometry/_core/angles.py (hunk 3 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/angles.py:378]: yes, you are right, i will incorporate that

File: src/compas/geometry/_core/angles.py (hunk 2 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/angles.py]: copied, and changed

File: src/compas/geometry/_core/angles.py (hunk 4 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/angles.py]: wasn't really aware of that, but makes total sense. Thank you very much for the input. Really appreciated.

File: src/compas/geometry/_core/intersections.py (hunk 5 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/intersections.py]: I will apply this to everything, thank you very much

File: src/compas/geometry/_core/intersections.py (hunk 6 above)
Comment (nikeftekhar) [reply src/compas/geometry/_core/intersections.py]: perfect, understood. i am learning a lot. Thank you very much. really appreciated.
